# Exploring Ebay Car Sales Data

In this guided project, we'll work with a dataset of used cars from eBay Kleinanzeigen,a [classifieds](https://en.wikipedia.org/wiki/Classified_advertising) section of the German eBay website.


The dataset was originally [scraped](https://en.wikipedia.org/wiki/Web_scraping) and uploaded to Kaggle by user [orgesleka](https://www.kaggle.com/orgesleka).The original dataset isn't available on Kaggle anymore,but can be found [here](https://data.world/data-society/used-cars-data).


***The aim of this project is to clean the data and analyze the included used car listings.***


A few modifications have been made to the dataset:
1. The data has been sampled to 50,000 data points from the full dataset as a sample.
2. The dataset has been modified to closely resemble to a scraped dataset(the version on kaggle is cleaned to be easier to work with).

The data dictionary provided with data is as follows:

| Column               	| Description                                                                   	|
|----------------------	|-------------------------------------------------------------------------------	|
| dateCrawled          	|  When this ad was first crawled.   All field values are taken from this date. 	|
| name                 	|  Name of the car.                                                             	|
| seller               	|  Whether the seller is private or a   dealer.                                 	|
| offerType            	|  The type of listing                                                          	|
| price                	|  The price on the ad to sell the   car.                                       	|
| abtest               	|  Whether the listing is included in   an A/B test.                            	|
| vehicleType          	|  The vehicle Type.                                                            	|
| yearOfRegistration   	|  The year in which the car was   first registered.                            	|
| gearbox              	|  The transmission type.                                                       	|
| powerPS              	|  The power of the car in PS.                                                  	|
| model                	|  The car model name.                                                          	|
| kilometer            	|  How many kilometers the car has   driven.                                    	|
| monthOfRegistration  	|  The month in which the car was   first registered.                           	|
| fuelType             	|  What type of fuel the car uses.                                              	|
| brand                	|  The brand of the car.                                                        	|
| notRepairedDamage    	|  If the car has a damage which is   not yet repaired.                         	|
| dateCreated          	|  The date on which the eBay listing   was created.                            	|
| nrOfPictures         	|  The number of pictures in the ad.                                            	|
| postalCode           	|  The postal code for the location   of the vehicle.                           	|
| lastSeenOnline       	|  When the crawler saw this ad last   online.                                  	|




In [1]:
#Import pandas and numpy library
import pandas as pd
import numpy as np
#Read the csv file.
#If encoding is not specified by default it takes it as UTF-8.
#It is throwing an error 
autos=pd.read_csv("autos.csv",encoding="Latin-1")



In [2]:
#After running the below cell jupyter will by default return first and last few rows of the dataset
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [3]:
#Information about the variables in the data
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

In [4]:
#First few rows of the dataset
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

After observing the dataset we have the following observations:
1. As mentioned at the start the dataset has 20 columns and 50000 columns most of which are string.
2. If we look at the information we can see that most of the columns do have any Null Values except for five.
3. The names of the columns are in [camelcase](https://en.wikipedia.org/wiki/Camel_case) instead of [snakecase](https://en.wikipedia.org/wiki/Snake_case).



In [5]:
#Print an array of the existing columns

print("The names of the existing columns are:")
autos.columns

The names of the existing columns are:


Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

## Modifying column names
We will make some edits to the column names:
1. Change the case of columns from camelcase to snakecase.
2. Change the names of some columns so that they become more descriptive.

The reason we will be doing this is to understand better the data actually contained in the respective columns.Also it will make easier when we use some columns in our further analysis

In [6]:
#Assign the original names of the column to an object
org_col_names=autos.columns
#Modify  the column names
mod_col_names=['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'pictures_in_ad', 'postal_code',
       'last_seen_online']
#Assigning the modified names to the original names
autos.columns=mod_col_names
#Check the initial rows of the data after modifying the names of the columns
print("Data after modifying the column names:")

autos.head()



Data after modifying the column names:


date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  pictures_in_ad  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code     last_seen_online  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

# Exploring the data
Now we will be doing some basic data exploration to determine what data cleaning needs to be done further.

* Look for the columns with string values and the values contained in it
* Look for columns which are stored as string but should be converted to numeric so that they can be used in a better way for further analysis


In [7]:
#Descriptive statistics for the data

print("Descriptive stats for data:")
autos.describe(include="all") 
      #include=all to get stats for categorical columns as well





Descriptive stats for data:


date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-09 11:54:38  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  pictures_in_ad   postal_code     last_seen_online  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

From the above output we can observe the below:
1. The columns seller,offer_type and unrepaired_damange majorly seem to have only one value and hence it might be redundant using them
2. The columns price and odometer are stored as strings when they shoud be numeric.
3. We can further look into the frequency distribution for gearbox and fuel_type since they seem to be dominated by one single value.
4. There seeme to be some discrepancy with the pictures_in_ad column since all metrics of meana and quartiles seem to be zero.
5. We also need to look into the registration year and registration month column since their minimum and maximum values seem to be bizarre.
6. For postal code we might need to consider converting it to categorical or string type because even though they are numbers they do no really mean that a higher postal code means a better place.
7. We might need to look a the data columns (ex:date_crawled) because they seem to be stored as strings making it difficult to understand them quantitatively.



## Exploring *gearbox* and *fuel_type* columns

In [8]:

#Counts for columns gearbox and fuel type
gearbox_counts=autos["gearbox"].value_counts()
fuel_type_counts=autos["fuel_type"].value_counts()

print("Counts for gearbox column","\n",gearbox_counts)
print("Counts for fule_type column","\n",fuel_type_counts)


Counts for gearbox column 
 manuell      36993
automatik    10327
Name: gearbox, dtype: int64
Counts for fule_type column 
 benzin     30107
diesel     14567
lpg          691
cng           75
hybrid        37
andere        22
elektro       19
Name: fuel_type, dtype: int64


After looking at the above output we can see that gearbox has mostly the value as *manuell* and fuel type has it as *benzin* and *diesel*.


## Cleaning and Converting *price* and *odometer* columns to numeric

In [9]:
#Converting the price column to numeric
#It has $ sign and ,.We need to replace those first and then convert it to numeric

autos["price"]=(autos["price"]
                .astype(str) #For some reason price is taken as int instead of str so first cast it as string
                .str.replace("$","")
                .str.replace(",","")
                .astype(int))

#Converting the odometer column to numeric
#It has km and ,.We need to replace those first and then convert it to numeric

autos["odometer"]=(autos["odometer"]
                   .astype(str)
                   .str.replace("km","")
                   .str.replace(",","")
                   .astype(int))

#Rename the odometer column to odometer_km
#This is done so that we do not miss out that the readings are in km
autos.rename({"odometer":"odometer_km"},axis=1,inplace=True)

autos.columns


Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer_km', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'pictures_in_ad', 'postal_code',
       'last_seen_online'],
      dtype='object')

## Exploring odometer_km and price columns
Now that we have changed the datatype from string  to numeric for price and odometer let's explore it even further.Let's try to look for distribution of values for those columns and detect outliers if any.

In [10]:
#Exploration of odometer_km column

print("The number of unique values in odometer_km are","",
      autos["odometer_km"].unique().shape[0])

print("\n","Odometer_km_Data statistics","\n",
      autos["odometer_km"].describe())
print("\n","Odometer_km_Value_Counts","\n",
      autos["odometer_km"].value_counts())




The number of unique values in odometer_km are  13

 Odometer_km_Data statistics 
 count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

 Odometer_km_Value_Counts 
 150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
5000        967
40000       819
30000       789
20000       784
10000       264
Name: odometer_km, dtype: int64


Looking at the odometer columns we can see that there are no outliers as such.However the concentration of values is more toward high mileage values than low mileage values

In [11]:
#Exploration of price column

print("The number of unique values in price are","",
      autos["price"].unique().shape[0])

print("\n","price_Data statistics","\n",
      autos["price"].describe())
print("\n","Lowest 20 prices with counts","\n",
      autos["price"].value_counts()
      .sort_index(ascending=True) #sort asc to sort low to high prices
      .head(20)) #head to get the top 10 values only

print("\n","Highest 20 prices with counts","\n",
      autos["price"].value_counts()
      .sort_index(ascending=False)
      .head(20))



The number of unique values in price are  2357

 price_Data statistics 
 count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

 Lowest 20 prices with counts 
 0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
20       4
25       5
29       1
30       7
35       1
Name: price, dtype: int64

 Highest 20 prices with counts 
 99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
Name: price, dtype: int64


While looking at the price column we can see that there are 1421 values with 0 as price which will be ~3% of the total values.Also when look at the higher price points we can see some cars with price greater than or nearby \$1 million.The count is 14.
After $999990 the value drops to \$350000.Hence we will retain data where price is between \$1 and \$350000 and drop the rest.

In [12]:
#exclude outliers
autos=autos[autos["price"].between(1,350000)]
#Data statistics after removing outliers
autos["price"].describe()

count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price, dtype: float64

## Exploring the date columns
There are 5 columns that represent date mainly date_crawled, last_seen,ad_created,registration date and registration year.
The date_crawled, last_seen, and ad_created columns are all sttored as string values. Because these three columns are represented as strings, we need to convert the data into a numerical form.The other two columns are represented as numeric values.

In [13]:
#Explore date_crawled, ad_created, and last_seen_online columns
print(autos["date_crawled"].head())
print(autos["ad_created"].head())
print(autos["last_seen_online"].head())

0    2016-03-26 17:47:46
1    2016-04-04 13:38:56
2    2016-03-26 18:57:24
3    2016-03-12 16:58:10
4    2016-04-01 14:38:50
Name: date_crawled, dtype: object
0    2016-03-26 00:00:00
1    2016-04-04 00:00:00
2    2016-03-26 00:00:00
3    2016-03-12 00:00:00
4    2016-04-01 00:00:00
Name: ad_created, dtype: object
0    2016-04-06 06:45:54
1    2016-04-06 14:45:08
2    2016-04-06 20:15:37
3    2016-03-15 03:16:28
4    2016-04-01 14:38:50
Name: last_seen_online, dtype: object


When we look at the above values we can see that the first 10 characters for all the three columns is date.We can extract the same and then calculate the distribution of the values.

In [14]:
#extract the date from columns and look at the distribution 
print("Bottom 20 dates and % share","\n",
      (autos["date_crawled"].str[:10]
 .value_counts(normalize=True,dropna=False) #normalize to get percentages
 .sort_index(ascending=True)
 .head(20)))
    
print("Top 20 dates and % share","\n",
      (autos["date_crawled"].str[:10]
 .value_counts(normalize=True,dropna=False) 
 .sort_index(ascending=False)
 .head(20)))
    
print("Top 20 dates as per highest % share","\n",
      (autos["date_crawled"].str[:10]
 .value_counts(normalize=True,dropna=False) 
 .sort_values(ascending=False)
 .head(20)))



    


Bottom 20 dates and % share 
 2016-03-05    0.025327
2016-03-06    0.014043
2016-03-07    0.036014
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-10    0.032184
2016-03-11    0.032575
2016-03-12    0.036920
2016-03-13    0.015670
2016-03-14    0.036549
2016-03-15    0.034284
2016-03-16    0.029610
2016-03-17    0.031628
2016-03-18    0.012911
2016-03-19    0.034778
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-22    0.032987
2016-03-23    0.032225
2016-03-24    0.029342
Name: date_crawled, dtype: float64
Top 20 dates and % share 
 2016-04-07    0.001400
2016-04-06    0.003171
2016-04-05    0.013096
2016-04-04    0.036487
2016-04-03    0.038608
2016-04-02    0.035478
2016-04-01    0.033687
2016-03-31    0.031834
2016-03-30    0.033687
2016-03-29    0.034099
2016-03-28    0.034860
2016-03-27    0.031092
2016-03-26    0.032204
2016-03-25    0.031607
2016-03-24    0.029342
2016-03-23    0.032225
2016-03-22    0.032987
2016-03-21    0.037373
2016-03-20    0.037887
2016-03-19 

From the above we can conclude that the site was almost daily crawled in the months of March and April.However the distribution through days seems to be uniform.

In [15]:
#extract the date from columns and look at the distribution - ad created
print("Bottom 20 dates and % share","\n",
      (autos["ad_created"].str[:10]
 .value_counts(normalize=True,dropna=False) #normalize to get percentages
 .sort_index(ascending=True)
 .head(20)))
    
print("Top 20 dates and % share","\n",
      (autos["ad_created"].str[:10]
 .value_counts(normalize=True,dropna=False) 
 .sort_index(ascending=False)
 .head(20)))
    
print("Top 20 dates as per highest % share","\n",
      (autos["ad_created"].str[:10]
 .value_counts(normalize=True,dropna=False) 
 .sort_values(ascending=False)
 .head(20)))



    


Bottom 20 dates and % share 
 2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
2015-12-30    0.000021
2016-01-03    0.000021
2016-01-07    0.000021
2016-01-10    0.000041
2016-01-13    0.000021
2016-01-14    0.000021
2016-01-16    0.000021
2016-01-22    0.000021
2016-01-27    0.000062
2016-01-29    0.000021
2016-02-01    0.000021
2016-02-02    0.000041
2016-02-05    0.000041
2016-02-07    0.000021
2016-02-08    0.000021
Name: ad_created, dtype: float64
Top 20 dates and % share 
 2016-04-07    0.001256
2016-04-06    0.003253
2016-04-05    0.011819
2016-04-04    0.036858
2016-04-03    0.038855
2016-04-02    0.035149
2016-04-01    0.033687
2016-03-31    0.031875
2016-03-30    0.033501
2016-03-29    0.034037
2016-03-28    0.034984
2016-03-27    0.030989
2016-03-26    0.032266
2016-03-25    0.031751
2016-03-24    0.029280
2016-03-23    0.032060
2016-03-22    0.032801
2016-03-21    0.037579
2016-03-20    0.037949
2016-03-19   

There is a large variety of dates in the ad created columns.However most of them seem to be in the 2 months of March and April 2016 distributed uniformly thorughout days.But there are also dates from long time back of 2015 but are very less in number.

In [16]:
#extract the date from columns and look at the distribution - last seen online
print("Bottom 20 dates and % share","\n",
      (autos["last_seen_online"].str[:10]
 .value_counts(normalize=True,dropna=False) #normalize to get percentages
 .sort_index(ascending=True)
 .head(20)))
    
print("Top 20 dates and % share","\n",
      (autos["last_seen_online"].str[:10]
 .value_counts(normalize=True,dropna=False) 
 .sort_index(ascending=False)
 .head(20)))
    
print("Top 20 dates as per highest % share","\n",
      (autos["last_seen_online"].str[:10]
 .value_counts(normalize=True,dropna=False) 
 .sort_values(ascending=False)
 .head(20)))



Bottom 20 dates and % share 
 2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-08    0.007413
2016-03-09    0.009595
2016-03-10    0.010666
2016-03-11    0.012375
2016-03-12    0.023783
2016-03-13    0.008895
2016-03-14    0.012602
2016-03-15    0.015876
2016-03-16    0.016452
2016-03-17    0.028086
2016-03-18    0.007351
2016-03-19    0.015834
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-22    0.021373
2016-03-23    0.018532
2016-03-24    0.019767
Name: last_seen_online, dtype: float64
Top 20 dates and % share 
 2016-04-07    0.131947
2016-04-06    0.221806
2016-04-05    0.124761
2016-04-04    0.024483
2016-04-03    0.025203
2016-04-02    0.024915
2016-04-01    0.022794
2016-03-31    0.023783
2016-03-30    0.024771
2016-03-29    0.022341
2016-03-28    0.020859
2016-03-27    0.015649
2016-03-26    0.016802
2016-03-25    0.019211
2016-03-24    0.019767
2016-03-23    0.018532
2016-03-22    0.021373
2016-03-21    0.020632
2016-03-20    0.020653
2016-03

The last seen online column marks the last time the crawler saw the listing meaning that the ad was taken out after that date maybe because it was sold.
However when we check we can see that the 5,6,7 dates of April seem to have a very high percentage share,almost 5 times higher compared to the previous days.
This means that the values majorly indicate the ending of crawling period and not necessarily car sales.


In [17]:
#Exploring registration year column
autos["registration_year"].describe()


count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

As we had already seen, the minimum and maximum values for *registration_year* column are 1000 and 9999 respectively.
Both the values are obviously incorrect.
Because a car can't be first registered after the listing was seen,any vehicle with a registration year above 2016 is definitely inaccurate.Determining the earliest valid year is more difficult.Realistically it could be somewhere in the first few decades of the 1900s.

We will try to look at the number of rows if we exclude values higher than 2016 and lesser than 1900 and check if it is safe to remove these rows.

In [18]:
#Check the values less than 200 and greater than 2016 for reg year column

autos["registration_year"].between(1900,2016).value_counts()


True     46681
False     1884
Name: registration_year, dtype: int64

From the above we can see that ~4% of the values have values beyond our specified bounds for registration_column.Hence we can safely remove those rows. 

In [19]:
#Remove the values less than 200 and greater than 2016 for reg year column
autos=autos[autos["registration_year"].between(1900,2016)]
print("Bottom 20 years","\n",
      (autos["registration_year"]
.value_counts(normalize=True)
.sort_values(ascending=True)
.head(20)))

print("Top 20 years","\n",
      (autos["registration_year"]
.value_counts(normalize=True)
.sort_values(ascending=False)
.head(20)))




Bottom 20 years 
 1952    0.000021
1953    0.000021
1943    0.000021
1929    0.000021
1931    0.000021
1938    0.000021
1948    0.000021
1927    0.000021
1939    0.000021
1955    0.000043
1957    0.000043
1934    0.000043
1951    0.000043
1941    0.000043
1954    0.000043
1950    0.000064
1962    0.000086
1937    0.000086
1958    0.000086
1956    0.000086
Name: registration_year, dtype: float64
Top 20 years 
 2000    0.067608
2005    0.062895
1999    0.062060
2004    0.057904
2003    0.057818
2006    0.057197
2001    0.056468
2002    0.053255
1998    0.050620
2007    0.048778
2008    0.047450
2009    0.044665
1997    0.041794
2011    0.034768
2010    0.034040
1996    0.029412
2012    0.028063
1995    0.026285
2016    0.026135
2013    0.017202
Name: registration_year, dtype: float64


From the above we can see that maximum registrations were made in the last 20 years.

# Analyzing variations across different car brands

In [20]:
#Analyzing the values in the brand column
unique_brands=autos["brand"].unique()
print("The number of unique brands is ",len(unique_brands))

#Look for top 20 brands with respect to percentage share in the data
top_10=autos["brand"].value_counts(normalize=True).sort_values(ascending=False).head(10)
print("The top 10 brands are:","\n",top_10)



The number of unique brands is  40
The top 10 brands are: 
 volkswagen       0.211264
bmw              0.110045
opel             0.107581
mercedes_benz    0.096463
audi             0.086566
ford             0.069900
renault          0.047150
peugeot          0.029841
fiat             0.025642
seat             0.018273
Name: brand, dtype: float64


## Analyzing prices of top brands

In [21]:
#Assign the names of top 20 brands to a variable
top_10_brands=top_10.index
#Create a dictionary to get mean price for every brand
brands_mean_price={}
for brand in top_10_brands:
    selected_rows=autos[autos["brand"]==brand]
    mean_price=selected_rows["price"].mean()
    brands_mean_price[brand]=int(mean_price)
    
print(brands_mean_price)


{'volkswagen': 5402, 'bmw': 8332, 'opel': 2975, 'mercedes_benz': 8628, 'audi': 9336, 'ford': 3749, 'renault': 2474, 'peugeot': 3094, 'fiat': 2813, 'seat': 4397}


Looking ath the average price we can see that:
1. There is a significant difference in the pricing for the top 10 brands in the dataset.
2. audi is the highest priced brand followed by mercedes_benz and bmw.
3. The brand opel,fiat,peugeot,ford and renault seem to be slightly on a lower price side compared to others.
4. Volkswagen being the top brand seem to lie somewhere in between high and low prices.It might be this is the reason for its popularity because it target both type of audiences.


## Analyzing mileage and prices of top brands

Now that we have looked into the prices let's look if the prices are connected with the mileage for the top 10 brands.We need to analyze both price and mileage.However comparing two series objects is slightly difficult.Hence we will be combining them in a single dataframe.


In [22]:
#Create a dictionary to get mean mileage for every brand
brands_mean_mileage={}
for brand in top_10_brands:
    selected_rows=autos[autos["brand"]==brand]
    mean_mileage=selected_rows["odometer_km"].mean()
    brands_mean_mileage[brand]=int(mean_mileage)
    
print(brands_mean_mileage)

{'volkswagen': 128707, 'bmw': 132572, 'opel': 129310, 'mercedes_benz': 130788, 'audi': 129157, 'ford': 124266, 'renault': 128071, 'peugeot': 127153, 'fiat': 117121, 'seat': 121131}


In [23]:
# Create a series from the two dictionaries
bp_series=pd.Series(brands_mean_price)
bm_series=pd.Series(brands_mean_mileage)

#Create a dataframe from first series 
df=pd.DataFrame(bp_series,columns=["mean_price"])
df

#Append column for second series 
df["mean_mileage"]=bm_series
df


mean_price  mean_mileage
volkswagen           5402        128707
bmw                  8332        132572
opel                 2975        129310
mercedes_benz        8628        130788
audi                 9336        129157
ford                 3749        124266
renault              2474        128071
peugeot              3094        127153
fiat                 2813        117121
seat                 4397        121131

As we can see from the above table there is not so much of a higher variation in mileage as it is in prices for top brands.
However we can see that there is a trend that higher price point vehicles give higher mileage compared to lower price point vehicles.